src/domain/core/src/repos/metadata_chain_visitors.rs

```diff
@@ -0,0 +1,402 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::error::Error;
+use std::marker::PhantomData;
+
+use internal_error::InternalError;
+use opendatafabric::{
+    AddData,
+    AsTypedBlock,
+    ExecuteTransform,
+    IntoDataStreamBlock,
+    MetadataBlock,
+    MetadataBlockDataStream,
+    MetadataBlockTyped,
+    MetadataEvent,
+    MetadataEventDataStream,
+    MetadataEventTypeFlags as Flag,
+    Multihash,
+    Seed,
+    SetAttachments,
+    SetDataSchema,
+    SetInfo,
+    SetLicense,
+    SetPollingSource,
+    SetTransform,
+    SetVocab,
+    VariantOf,
+};
+
+use crate::{HashedMetadataBlockRef, MetadataChainVisitor, MetadataVisitorDecision as Decision};
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct SearchSetVocabVisitor<E = InternalError> {
```

Comment:
Ohh, I think I finally get it ... it's because when we accept multiple visitors they ALL have to agree on a single error type. That's an interesting problem.

And I think I understand why you picked `InternalError` as the common denominator.

Reply:
So if we have visitors A and B that **agree** on error types - below just works:
```rust
let visitor_a = VisitorA::new();
let visitor_b = VisitorB::new();
chain.accept(&mut[
  &mut visitor_a,
  &mut visitor_b,
]).await?;
````

If they **disagree** your current approach is to parametrize them with some common generic error type that their "true" error type converts into.

I think a better solution to this problem is (once again) composition:
```rust
let visitor_a = VisitorA::new().map_err(|e| ...);
let visitor_b = VisitorB::new().map_err(|e| ...);
chain.accept(&mut[
  &mut visitor_a,
  &mut visitor_b,
]).await?;
````
Where `map_err()` is implemented for `MetadataChainVisitor` trait and wraps visitors with a simple struct that converts its error.

Reply:
Second problem I see is what error should visitors that aren't expected to fail return...

In future this might be a good case for never type (`!`) but its not yet stabilized for such general-purpose use...

We could introduce a `Never` error type specifically for the visitor case (that can be used but never created).

Reply:
> In future this might be a good case for never type (!) but its not yet stabilized for such general-purpose use...

Yes, indeed -- I considered this option, but Rust in this case reduces the error type to the maximum, since it expects the never type everywhere. maybe I'm spoiled by the never type in Haskell, where the situation is the opposite (if I'm not confused).

Reply:
> Ohh, I think I finally get it ... it's because when we accept multiple visitors they ALL have to agree on a single error type. That's an interesting problem.

Yeah, it's exactly like that -- mentioned it in a thread above (https://github.com/kamu-data/kamu-cli/pull/543#discussion_r1538388389)

Reply:
> I think a better solution to this problem is (once again) composition:

Interesting point, I have my doubts about the error type output -- I'll have to take an extra portion of programmer's fuel (coffee) to implement that

Reply:
Done